bloom-instance/dns/zone-resolver/main.tf

```diff
@@ -0,0 +1,40 @@
+
+# This module doesn't create any resources.  All it does is provide a simple, 
+# repeatable means for determining which zone ID to use for applying DNS records
```

Comment:
is there a way to write tests for the following logic? What are the alternatives to not automating the selection of zone per domain- is it explicitly adding a domain to a config file?

Reply:
For the cert code, these lookups simply the process of adding certs without having to explicitly pass a zone ID (or zone name if we enable direct lookups against the `dns` config).  That's simple enough to fall back on, if a bit more verbose and error prone, especially when copying and pasting between different tfvars files.

The main area where these lookups comes in is in dynamically mapping service domain names to the ALB endpoint via CNAME records.  Right now that's using lookup logic in the `dns/record` module that was stripped out of the original `dns/zone-resolver` module, but that will probably be modified again to use the `zone-resolver` module instead since the logic is a bit more robust.  Adding zone names/IDs for those records becomes a bit more complicated, since they are based on the knowledge that domain ABC routes ALB requests to service XYZ.  Once I add in the CloudFront config, it would probably be even trickier.

Terraform is very limited in terms of proper testing.  You can't even add print statements, much less assert.  Unfortunately, Terraform [doesn't have the ability to define custom functions](https://github.com/hashicorp/terraform/issues/27696), either, so modules are how we approximate that functionality.  One thing that we might be able to do is create a separate `modules` directory outside of the `bloom-instance` project that holds both the module code and a full Terraform config.  The module could be loaded dynamically from wherever, but the project itself could define a set of module resources with inputs and expected outputs.  The output vars for the project could contain comparison expressions evaluating whether the passed values resulted in the expected values.  It's a bit hacky and I don't see a good way to automate that, but it would at least provide a testbed for validating module output.

Reply:
There are 3 possibilities:

1) No matching zone is found, which will result in an error
2) The wrong zone is matched, and it isn't a valid root domain (ie creating records for *.example.com in the example.org zone). I'm pretty sure this will also result in an error on the AWS side, but I haven't tested.
3) The wrong zone is matched, but it is a valid root domain (ie creating records for *.dev.example.com in the example.com zone rather than dev.example.com). These records will create without error, but will not be resolvable since the correct zone (dev.example.com) doesn't have them. This will just cause automatic cert validation to fail, which will be visible during deployment.

Reply:
Note that 2 above is extremely unlikely unless there are significant changes made to the matching logic.  The very first step filters out any domains that don't have a shared prefix, so there's no way for `example.com` to match `example.org`.  The other steps are designed to prevent 3 by using the longest match rather than the first one.  That means that `api.dev.example.com` will select `dev.example.com` even if `example.com `is also a match.  Those steps are a bit more complex, so if anything unexpected were to happen, that's probably where it would be.